ratelimit_mock_time_test.go

```diff
@@ -0,0 +1,111 @@
+package ratelimit
+
+/**
+This fake time implementation is a modification of time mocking
+the mechanism used by Ian Lance Taylor in https://github.com/golang/time project
+https://github.com/golang/time/commit/579cf78fd858857c0d766e0d63eb2b0ccf29f436
+
+Modified parts:
+ - advanceUnlocked method uses in-place filtering of timers
+   instead of a full copy on every remove.
+   Since we have 100s of timers in our tests current linear
+   the complexity of this operation is OK
+   If going to have 1000s in the future, we can use heap to store timers.
+ - advanceUnlocked method yields the processor, after every timer triggering,
+   allowing other goroutines to run
+*/
+
+import (
+	"runtime"
+	"sync"
+	"time"
+)
+
+// testTime is a fake time used for testing.
+type testTime struct {
+	mu     sync.Mutex
+	cur    time.Time   // current fake time
+	timers []testTimer // fake timers
+}
+
+// makeTestTime hooks the testTimer into the package.
+func makeTestTime() *testTime {
+	return &testTime{
+		cur: time.Now(),
+	}
+}
+
+// testTimer is a fake timer.
+type testTimer struct {
+	when time.Time
+	ch   chan<- time.Time
+}
+
+// now returns the current fake time.
+func (tt *testTime) now() time.Time {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	return tt.cur
+}
+
+// newTimer creates a fake timer. It returns the channel,
+// a function to stop the timer (which we don't care about),
+// and a function to advance to the next timer.
+func (tt *testTime) newTimer(dur time.Duration) (<-chan time.Time, func() bool, func()) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	ch := make(chan time.Time, 1)
+	timer := testTimer{
+		when: tt.cur.Add(dur),
+		ch:   ch,
+	}
+	tt.timers = append(tt.timers, timer)
+	return ch, func() bool { return true }, tt.advanceToTimer
+}
+
+// advance advances the fake time.
+func (tt *testTime) advance(dur time.Duration) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	tt.advanceUnlocked(dur)
+}
+
+// advanceUnlock advances the fake time, assuming it is already locked.
+func (tt *testTime) advanceUnlocked(dur time.Duration) {
+	tt.cur = tt.cur.Add(dur)
+
+	i := 0
+	j := 0
+	for i < len(tt.timers) {
+		if tt.timers[i].when.After(tt.cur) {
+			if i != j {
+				tt.timers[j] = tt.timers[i]
+			}
+			i++
+			j++
+		} else {
+			tt.timers[i].ch <- tt.cur
+			for i := 0; i < 16; i++ {
+				runtime.Gosched()
+			}
```

Comment:
This also not present in https://github.com/golang/time/commit/579cf78fd858857c0d766e0d63eb2b0ccf29f436.

I guess you're saying we need to gosched 16 times to get the desired consistency?

Reply:
We just run much more goroutines concurrently than golang/time tests, so we need to give them more time to run their logic.

Reply:
Weird, I wish there was a way to tell runtime "run everything else but me". And `time.Sleep(Nanosecond) is not great.

Ack.